rfd/XXXX-cloud-agent-upgrades.md

```diff
@@ -0,0 +1,318 @@
+down the provided windows to those whose `EndTime` is in the future, and whose `StartTime` is before `now + max_maintenance_interval`
+(likely 1.5-2 weeks). If at least one upgrade window exists post-filter, at least one attempt at discovery + restart aught
+to be made within one of the provided windows. Additionally, at least one discovery + restart attempt must be made at least
+every `max_maintenance_interval` regardless.
+
+The additional `AppearsOutdatedHint` flag will permit the auth server to suggest that the *specific* client version sent
+in the request is too old. By default, agents should ignore windows that started *before* the start time of
+the current process, and should endeavor to spread out restarts through semi-random window selection. The presence of
+this flag indicates that these behaviors should be overridden (likely due to a critical patch becoming available).
+
+Because the control plane might also restart during a maintenance window, direct contact with the auth server should
+not be a requirement to perform a restart. Agents should cache maintenance windows locally, only periodically
+refreshing their local state (though it should always be re-pulled on restart). Furthermore, agents that are randomly
+selecting a restart time should not recalculate it on pull unless the previously selected time no longer falls within
+a maintenance window, or the `AppearsOutdatedHint` flag was present and the previously selected time is not within
+the next/current window.
+
+
+### Kubernetes Model
+
+For kubernetes based agents, discovery and restart will be managed by a separate cloud-stable upgrade controller.
+In order to minimize the need for maintenance/upgrade of the controller itself, the controller will not have
+any direct connection to teleport cluster state. Instead, agents will periodically write out maintenance window
+information to a configuration resource accessible to the controller.
+
+The upgrade controller will be responsible for all teleport agents within its namespace, and will perform rolling
+updates, applying the latest immutable image tag to all agents. Updates will be applied if one of the following
+conditions is met:
+
+1. A new cloud-stable image is available and we are within an update window (checked on periodic tick).
+2. Some threshold of agents are in a CrashLoop (controller is woken up on pod state change).
+
+Because this model doesn't cause downtime unless an update is *actually* available (unlike the apt/yum model),
+it is more acceptable to always select the next available maintenance window for applying the update. In theory,
+it would still be best to only perform updates about once per week, but the need is less, so we'll likely use the
+simpler greedy approach, at least at first.
+
+Some of the details of the controller are still TBD, but the basic overview looks something like this:
```

Comment:
This schema represents my current comprehension of the updater's behaviour.

```mermaid
graph TD
    event([event])
    event --> potentialUpdate

    subgraph potentialUpdate
    isHealthy{Is Healthy ?}
    inMaintenance{In maintenance ?}
    canUpdate

    isHealthy -- no --> canUpdate
    isHealthy -- yes --> inMaintenance
    inMaintenance -- yes --> canUpdate
    inMaintenance -- unknown --> canUpdate
    
    end
    inMaintenance -- no --> requeue

    subgraph versionAvailable
    getVersion[Get version from X]
    compare{Version compare}
    isDowngradeLegal{Is downgrade ok?}

    getVersion --> compare
    compare -- lower --> isDowngradeLegal
    end

    canUpdate --> versionAvailable
    compare -- same --> requeue
    isDowngradeLegal -- no --> requeue

    subgraph imageCheck
    getManifest
    keylessSign{Keyless sign OK?}
    staticSign{Static sign OK?}

    getManifest -- ok --> keylessSign
    keylessSign -- no --> staticSign
    end
    getManifest -- fail --> requeue
    keylessSign -- yes --> updateDeployment
    staticSign -- yes --> updateDeployment
    staticSign -- no --> requeue

    compare -- higher --> imageCheck
    isDowngradeLegal -- yes --> imageCheck

    updateDeployment[Update Deployment]


    requeue[Requeue in 30min]
```

Reply:
Do we want to invert the `versionAvailable` and `potentialUpdate` blocks? This would cause more calls to the version server but would allow measuring how many updaters are running in the wild based on request rate.

Reply:
I wonder if we should do some caching/hysteresis to reduce spam against the version server.

Let's say I'm running 11.3.1 and the updater is woken up and updates to 11.3.2. The updater will receive several events regarding the deployment in the next seconds/minutes as the pods are gradually rolled-out. We don't want to hit the version server on all the subsequent reconciliations.

Reply:
Keyless/Static sign is discussed [in this comment](https://github.com/gravitational/teleport/pull/19045#discussion_r1101967322) on the image signing RFD.

Downgrade constraints are discussed [in this comment](https://github.com/gravitational/teleport/pull/20622#discussion_r1101974215).